src/libsyntax/attr/builtin.rs

```diff
@@ -592,81 +592,98 @@ fn find_deprecation_generic<'a, I>(sess: &ParseSess,
     let diagnostic = &sess.span_diagnostic;
 
     'outer: for attr in attrs_iter {
-        if attr.path != "deprecated" {
-            continue
+        if !attr.check_name("deprecated") {
+            continue;
         }
 
-        mark_used(attr);
-
         if depr.is_some() {
             span_err!(diagnostic, item_sp, E0550, "multiple deprecated attributes");
             break
         }
 
-        depr = if let Some(metas) = attr.meta_item_list() {
-            let get = |meta: &MetaItem, item: &mut Option<Symbol>| {
-                if item.is_some() {
-                    handle_errors(sess, meta.span, AttrError::MultipleItem(meta.name()));
-                    return false
-                }
-                if let Some(v) = meta.value_str() {
-                    *item = Some(v);
-                    true
-                } else {
-                    if let Some(lit) = meta.name_value_literal() {
+        depr = match &attr.meta().unwrap().node {
+            MetaItemKind::Word => Some(Deprecation { since: None, note: None }),
+            MetaItemKind::NameValue(lit) => {
+                match &lit.node {
+                    ast::LitKind::Str(s, ..) => {
+                        Some(Deprecation { since: None, note: Some(*s) })
+                    }
+                    _ => {
                         handle_errors(
                             sess,
                             lit.span,
                             AttrError::UnsupportedLiteral(
-                                "literal in `deprecated` \
-                                value must be a string",
-                                lit.node.is_bytestr()
+                                "deprecated reason must be a string literal",
```

Comment:
note to reviewer: this is dead code since the error gets caught much earlier by the attribute format check (see the test that I added). However, since I have to do the match anyways, I figured I might as well add the error handling code.

Reply:
`meta.value_str().map(|s|  Deprecation { since: None, note: Some(*s) })`
and you'll get rid of the whole match.

Reply:
Done.